Everyone pileon modifications to bolt12

rustyrussell opened the pull request:
Please leave feedback...

rustyrussell committed:
BOLT 12: fallout from compulsory paths.

1. final node_id in blinded path cannot be the real node_id, but it's still
   useful if you don't care about node_id.
2. blinded_capacities is redundant since htlc amounts are in blinded_payinfo.
3. Similarly, cltv.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: remove previously-optional signature from offers.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: simplify nonce calculation

Use first field, arrange for it to be `payer_info`.  This makes it
possible to calculate the merkle (in theory, at least), in a single
pass.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: mirror all offer fields into invoice_request, and similarly invoice.

This mainly involves renaming fields but also:
1. Renumber, so 2 - 127 are offers, 128-192 are invoice_request.  That
   makes for simply checks that offers aren't trying to sneak in new
   invoice_request fields etc.
2. Removes the concept of offer_id.
3. Add offer_metadata blob so you can do stateless offers more easily than
   having to insert data in a blinded path.
4. We do insert offer_node_id if it's implied by offer.
5. Feature fields have to be separate for each one.
6. We are no longer allowed to change description between offer and invoice
   (we can add a new "annotation" field if we wanted).

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

t-bast commented:
> blinded_capacities is redundant since htlc amounts are in blinded_payinfo.

What do you mean here? `blinded_payinfo` contains an `hltc_maximum_msat` field, but that's different from the path capacity. We could for example have a blinded path where:

- `htlc_maximum_msat` is set to 10 mbtc
- the available capacity is 50 mBTC
- the payment we expect is 25 mBTC

The sender should ideally send three HTLCs through this blinded path for the payment to succeed.
I'm in favor of adding the `capacity` field to `blinded_payinfo` directly, instead of using a separate tlv.

I'll let @thomash-acinq review the other changes and provide feedback.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -120,8 +120,8 @@ Each form is signed using one or more *signature TLV elements*: TLV
-`messagename` is the name of the TLV stream being signed (i.e. "offer", "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
-signature (e.g. "signature" or "refund_signature").
+`messagename` is the name of the TLV stream being signed (i.e. "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
+signature (e.g. "signature").
```
The example below still uses offer signature. Lines 143 and 174 in 4eb8b86a6429650b151c89914729b52d97d7af77.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -319,7 +311,7 @@ A reader of an offer:
 ## Rationale
 
-A signature is optional because it makes for a longer string (potentially
+A signature unnecessary, and makes for a longer string (potentially
```
s/unnecessary,/is unnecessary

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -328,6 +327,9 @@ invoices is `lnr`.
 2. types:
+    1. type: 1 (`payer_info`)
+    2. data:
+        * [`...*byte`:`blob`]
```
Is the choice of an odd field deliberate aside from making this always the first field if set?

Related is there any reason there are other odd fields (e.g. `chain`) at this point?

rustyrussell replied:
I was a bit nervous about using 0.  But TBH I should.  Will fix.

The only odd fields now are:

`payer_info`: (fixed to `invoice_payer_info`!) because it's actually valid to ignore this (you have to copy ignored fields anyway)
`invoice_payer_note`: same.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -186,43 +186,46 @@ The human-readable prefix for offers is `lno`.
     2. data:
         * [`...*chain_hash`:`chains`]
-    1. type: 6 (`currency`)
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
```
This is not described below.

rustyrussell replied:
Added "MAY set `offer_metadata` for its own use." just so it gets a mention, added Rationale.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -290,26 +294,32 @@ A writer of an offer:
 
 ## Rationale
 
+The entire offer is reflected in the invoice_request, both for
+completeness (so all information will be returned in the invoice), and
+so that the offer node can be stateless.
```
The commit description says:

> 3. Add offer_metadata blob so you can do stateless offers more easily than
>   having to insert data in a blinded path.

But here the offer reflection is stated as the reason for statelessness. Seems like reflecting fields adds more complexity as you now need to keep track of unknown fields when parsing instead of simply ignoring them.

If statelessness can be achieved via the `offer_metadata` field, that leaves "completeness" for the justification for reflecting the offer. What's the rationale for needing such completeness?

rustyrussell replied:
With full reflection, you only need the offer_metadata to contain the auth cookie (e.g SHA256(my-secret || offer)).  Without that, it also needs to duplicate the unreflected fields, making it potentially much larger.

Since we reflect many fields anyway, reflecting them all is actually simpler, and minimizes the offer size.

carlaKC commented on 12-offer-encoding.md:
```diff
@@ -362,6 +361,7 @@ invoices is `lnr`.
 ## Requirements for Invoice Requests
 
 The writer of an invoice_request:
+  - SHOULD set `payer_info` to an unpredictable series of bytes.
```
Why should over must? If we want "an unpredictable nonce for hashing" and we don't include `payer_info`, our `chain` tlv will be used, which is pretty predictable?

rustyrussell replied:
If you don't care about privacy?  Should we say MUST?

I set this to the data I need to recover the payer_key from my node_id, which (1) means I don't need to remember it, but can reconstruct from any invoice and (2) it's unpredictable to anyone else.

carlaKC replied:
I meant populating the `payer_info` in general, rather than the value that we have in it: 
- MUST set `payer_info` to a non-zero series of bytes
  - SHOULD set `payer_info` to an unpredictable series of bytes

Reasoning being that if we're going to use `first-tlv` in nonce calculation, not setting `payer_info` (which this wording allows) would mean you use the predictable `offer_chains` tlv?

carlaKC replied:
(somewhat relevant to [this comment](https://github.com/rustyrussell/lightning-rfc/pull/7/files#r948137786) about specifying using `invoice_request_payer_info` rather than just "hoping it's there")

carlaKC commented on 12-offer-encoding.md:
```diff
@@ -360,83 +405,88 @@ invoices is `lnr`.
-  - if the offer had a `quantity_min` or `quantity_max` field:
-    - MUST set `quantity`
+The writer:
+  - MUST copy all fields from the offer (including unknown fields).
```
I didn't quite follow the conversation in the spec meeting. What's the advantage of copying all the offer TLVs into the invoice request vs having a single `offer_tlvs` type which contains the full offer tlv stream (tlv-in-a-tlv)?

rustyrussell replied:
You can't partially prove that, you'd need to expose all of it.  And we were reflecting *many* fields already, so this is actually simpler.

jkczyz replied:
I suppose reflecting the fields is functionally equivalent to parsing two (or more) TLV streams back-to-back over a specific type range? (modulo `invoice_request_payer_info` which may be can be thought of as its own TLV stream?)

rustyrussell replied:
But simpler.  If you have two TLVs you need a wrapper format (they don't naturally separate, in general).  And you have, again, the "but we want a single proof" problem, which now has to handle two TLVs.

rustyrussell commented:
> > blinded_capacities is redundant since htlc amounts are in blinded_payinfo.
> 
> What do you mean here? `blinded_payinfo` contains an `hltc_maximum_msat` field, but that's different from the path capacity. We could for example have a blinded path where:
> 
>     * `htlc_maximum_msat` is set to 10 mbtc
> 
>     * the available capacity is 50 mBTC
> 
>     * the payment we expect is 25 mBTC
> 
> 
> The sender should ideally send three HTLCs through this blinded path for the payment to succeed. I'm in favor of adding the `capacity` field to `blinded_payinfo` directly, instead of using a separate tlv.

IMO, the purpose of htlc_max_msat is to limit HTLC exposure, it's very antisocial to use multiple HTLCs of this value at once.  I think the sender needs to set a value which is, if possible, a reflection of known capacity estimates, and also <= the htlc_max of any hop.  Keeps it to a simple "don't send more than this down this path".

rustyrussell committed:
Excellent feedback from @jkczyz

Only non-textual change is that invoice_payer_info is now type 0, not 1.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4, 9: use a different set of features for bolt12.

I lazily said "reuse bolt11 features", but all of them are currently
assumed or don't apply.  Create a new marker, currently unused.

Also make sure we *check* them everywhere, and clarify use of
`features` in `blinded_payinfo`.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: remove refund payer proof.

It's cool, but better to add later (esp since nobody has implemented it).
You can do a specific offer to a user for now.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: remove invoice replacement.

An obvious candidate for a new feature bit "I offer invoice replacement!".

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: renumber fields.

Now we've cleaned the fields, renumber consecutively into:

* offer: 1 - 79 incl
* invoice_request: 80 - 159 incl
* invoice: 160 - 239 incl

And clearly rename invoice_request fields to invoice_request_X.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -187,46 +185,43 @@ The human-readable prefix for offers is `lno`.
     2. data:
         * [`...*chain_hash`:`chains`]
-    1. type: 6 (`currency`)
+    1. type: 4 (`offer_metadata`)
```
Any reason `offer_metadata` isn't reflected in `invoice`?

rustyrussell replied:
Would you believe it was to check if anyone was paying attention?  No?  Anyone?

:)

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
 1. `tlv_stream`: `invoice_request`
 2. types:
-    1. type: 3 (`chain`)
+    1. type: 0 (`invoice_request_payer_info`)
```
Naming this `invoice_request_payer_info` may be confusing as it used differently in `invoice` (i.e., not copied from `invoice_request` IIUC). Should we special case the naming here?

rustyrussell replied:
Hmm, yeah.  If you have a send_invoice offer, you go straight to invoice, pay.  In this case, there's no payer key and payer info, unless we put that in the offer.  See comments below...

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -551,144 +582,100 @@ using the `onion_message` `invoice` field.
-      - MUST NOT set `refund_signature`
+      - MUST NOT set `invoice_request_quantity`
+    - MUST set `invoice_request_payer_key` to `offer_node_id`.
+    - MUST set `invoice_request_payer_info`
+	  - SHOULD set it to at least 16 random bytes.
```
Could we describe this more either here or in the rationale, assuming it differs from the same field in `invoice_request`?

thomash-acinq replied:
Actually why is it allowed to touch this field? Shouldn't it be copied as-is from the invoice request?

rustyrussell replied:
There's no invoice_request in this case (above it says:   - otherwise (responding to a `offer_send_invoice` offer):).

From a high level, an 'send_invoice' offer *is* an invoice request.  Now invoice_request *has all the offer fields*, do we just, y'know, use it directly instead of having this weird kind of offer?  I need to think through this...

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
 ## Rationale
 
-A signature is optional because it makes for a longer string (potentially
+The entire offer is reflected in the invoice_request, both for
+completeness (so all information will be returned in the invoice), and
+so that the offer node can be stateless.  This makes `offer_metadata`
+particularly useful, since it can contain an authentication cookie to
+validate the other fields.
```
I'm fine with it if that's what everyone wants but I think copying everything wastes message space (not so long ago you wanted to use point32 to shave one byte from node ids), and the `offer_metadata` will take even more space from the QR code.
A customer can request invoices which makes it a DoS vector and stateless invoices are a great way to protect against it, however offers should be created at the node operator request and it not a DoS vector. If a node creates offers when a customer requests it, I think it could probably be replaced by an invoice.

rustyrussell replied:
I wanted to shave a byte off the QR code, sure.  But invoice / invoice_request space is way less expensive?  And offer_metadata is really only for stateless invoices; I don't expect to use it myself.

Full reflection is actually a nice simplification, since we did per-field reflection before.  (Disclaimer: I haven't *implemented* it yet!).

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
+    1. type: 0 (`invoice_request_payer_info`)
     2. data:
-        * [`chain_hash`:`chain`]
-    1. type: 4 (`offer_id`)
```
I liked the `offer_id`, when I received an invoice request I could easily get the matching offer. Now I have to extract the offer fields and compute a hash myself before checking if I have such an offer.

rustyrussell replied:
Yes, me too.  However, the offer fields are contiguous, so you can simply hash from the end of `payer_info` to the first field >= 128.

jkczyz replied:
s/128/80 ?

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -120,18 +119,17 @@ Each form is signed using one or more *signature TLV elements*: TLV
 The Merkle tree's leaves are, in TLV-ascending order for each tlv:
 1. The H("LnLeaf",tlv).
-2. The H("LnAll"||all-tlvs,tlv) where "all-tlvs" consists of all non-signature TLV entries appended in ascending order.
+2. The H("LnNonce"||first-tlv,tlv-type) where first-tlv is the numerically-first TLV entry in the stream, and tlv-type is the "type" field (1-9 bytes) of the current tlv.
```
It's unclear to me how this change is an improvement, now instead of using the entropy of the whole TLV stream, we rely only on the first TLV which is quite random. It seems that the first TLV will always be `invoice_request_payer_info` but in that case we should write that we use `invoice_request_payer_info` instead of just "first-tlv" and hoping that it is always `invoice_request_payer_info`.

rustyrussell replied:
This was a debate with Matt.  Using the first one makes is streamable, and generic, and avoid violating layers.  If some stream does not arrange to have their nonce field first, they lose the private partial proof property, but, well, let's not do that!

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -706,23 +693,19 @@ Note that the recipient of the invoice can determine the expected
-It's often useful to provide capacity hints, particularly where more
-than one blinded path is included, for payers to use multi-part
-payments.
+The invoice issuer is allowed to ignore `invoice_request_payer_note` (it has an odd
+number, so is optional), but must still copy it exactly.
```
This is ambiguous, can I ignore the field (which to me means do as it wasn't there) or do I have to copy it?

rustyrussell replied:
This is the rationale section, the requirements are clear about this:

```
  - if creating an `invoice` for an offer with `offer_send_invoice`:
     - MUST copy all fields from the offer (including unknown fields).
```
But perhaps ignore is too strong here; there's already a section on this in the "invoice_request" rationale.

```
`invoice_request_payer_note` allows you to compliment, taunt, or otherwise engrave
graffiti into the invoice for all to see.
```

I'll remove this.

thomash-acinq reviewed (commented):
I don't have objections to these changes but I'm not convinced that they improve anything either.

valentinewallace commented on 09-features.md:
```diff
@@ -25,6 +25,7 @@ The Context column decodes as follows:
 * `C-`: presented in the `channel_announcement` message, but always odd (optional).
 * `C+`: presented in the `channel_announcement` message, but always even (required).
 * `9`: presented in [BOLT 11](11-payment-encoding.md) invoices.
+* `2`: presented in [BOLT 12](12-offer-encoding.md) messages.
```
Is MPP support assumed in offers? LDK likely wants to support phantom invoices for offers, which don't support MPP..

rustyrussell replied:
I'm not sure what phantom invoices are, but yes, MPP is assumed?

valentinewallace replied:
Ah, LDK will need MPP support to be optional in `invoice` OMs at least.. I'm not sure the best way of doing this, maybe it would be easiest to put the MPP feature bit in offer features (though it has a dependency on the `payment_secret` feature currently, which is kinda awkward)?

rustyrussell replied:
Wait, how do you *not* support MPP?  In 2022, that seems like more work than supporting it everywhere (we don't *send* MPP for keysend, but of course we have no problem receiving it...).

valentinewallace replied:
Heh, we support MPP for regular invoices. But for multi-node receive invoices where the payment can be received by any backing node, we don't support MPP (because the nodes would have to coordinate with each other before releasing the preimage, which is not supported).

rustyrussell replied:
Sorry, that's not valid :(

You have an implementation (or maybe even a design?) which doesn't work for modern payments.  That you can currently make it work at all is a historical accident.  But really, not something we want to be designing around!

TheBlueMatt replied:
I don't buy this at all - its not really all that clear that MPP is really important for small-medium payments today, and skipping it can be *very* useful for small-medium payments today, as Val points out. For larger payments, indeed, its definitely required, but that's a separate issue (and actually, offers is really great there - the recipient can change the features based on the payment size).

TheBlueMatt replied:
Frankly, if we can't get an MPP feature bit we'll probably not realistically ship BOLT12, cause phantom invoices is a pretty useful feature for large scale deployments.

thomash-acinq replied:
It seems that what you want is to force the sender to send everything to the same blinded path. Disabling MPP completely is a crude way of doing it, the sender could still split the payment as long as all the part have the same destination.
If you want to enable a new design, you should add a new feature for it, not disable an existing feature that's considered part of the base package.

TheBlueMatt replied:
I'm not sure who determines what is a "part of the base package", but it's not clear to me MPP provides any value for small-medium payments. Indeed, eventually the solution to enable seamless failover without trusting a coordinator would be trampoline, but in the mean time things don't seem to break.

jkczyz replied:
> It seems that what you want is to force the sender to send everything to the same blinded path.

FYI, see the following for full context: https://lightningdevkit.org/blog/introducing-phantom-node-payments/

The key point is being able to retry on a different path (i.e., failover as @TheBlueMatt mentioned) without coordination between nodes.

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -669,13 +676,18 @@ A reader of an invoice:
   - MUST reject the invoice if `invoice_paths` is not present or is empty.
   - MUST reject the invoice if `invoice_blindedpay` is not present.
-    - MUST reject the invoice if `invoice_blindedpay` does not contain exactly one `blinded_payinfo` per `invoice_paths`.`blinded_path`.
+  - MUST reject the invoice if `invoice_blindedpay` does not contain exactly one `blinded_payinfo` per `invoice_paths`.`blinded_path`.
```
s/.//

Edit: oh, it looks like this is pseudocode-y syntax and everywhere, never mind lol

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
```
I'm wondering if we need to mirror this particular field in `invoice_request`s and `invoice`s? It seems like the biggest potential bandwidth consumer in the mirroring, and the other mirrored fields seem more relevant to determining whether an incoming `inv_req` is valid 🤔

rustyrussell replied:
Sure, but simple rules are good.  I mean, we could have a non-mirrored range, too.

I suppose a really dumb implementation could reuse these blinded paths for the invoice (they have to decrypt and re-encrypt them with more fields, which I *think* is possible if you store the blinding secret (or enough to derive it) in offer_metadata?).

valentinewallace replied:
Not sure I followed how these could be reused for the invoice, but I'm good with mirroring all fields for simplicity :)

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -706,23 +693,19 @@ Note that the recipient of the invoice can determine the expected
-It's often useful to provide capacity hints, particularly where more
-than one blinded path is included, for payers to use multi-part
-payments.
+The invoice issuer is allowed to ignore `invoice_request_payer_note` (it has an odd
+number, so is optional), but must still copy it exactly.
```
I'm a little confused by the use of an odd number for `invoice_request_payer_note`. I thought odd meant it is ok if you don't know how to interpret this type (i.e., the implementation predates this type). Not this is optional.

rustyrussell replied:
Yes, language here was too loose.  It's "optional to understand"...

rustyrussell committed:
BOLT 12: even more feedback and fixes.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell commented:
> I don't have objections to these changes but I'm not convinced that they improve anything either.

But we finally got rid of the "LnAll" string constant you hated!

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,335 @@ The human-readable prefix for offers is `lno`.
+  - MUST copy all fields from the offer (including unknown fields).
+  - MUST NOT set any tlv fields greater or equal to 160.
+  - if `offer_node_id` is not present:
+     - MUST set `offer_node_id` to the final `onionmsg_path` `point` in the first `offer_blinded_path`
```
I'm fine with copying the offer as-is but if we start modifying fields in the offer part, then we can no longer extract the offer from the invoice request and it becomes really messy.

rustyrussell replied:
Let's remove this optimization?  The point is that the offer_node_id may be completely made up.  But I agree this makes life messier.

thomash-acinq replied:
I think the optimization is still useful, blinded routes are probably what takes the most space so any optimization there is welcome. We should just not modify `offer_node_id` in the invoice request, we can just do it implicitly as we already do when reading an offer, no need to make it explicit.

rustyrussell replied:
The problem is that we use "no offer_node_id" as indicating that there was no offer (it's by far the cleanest solution).  I would re-add this optimization later, TBH.

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
-    - if the offer had a `quantity_min` or `quantity_max` field:
-      - MUST set `quantity`
+    - if `invoice_request_amount` is not present:
+      - MUST set `invoice_request_amount`.`msat` to the *base invoice amount* calculated from the invoice_request
```
Again, I think we should not touch the invoice request part when creating the invoice. Just add a new field `invoice_amount`.

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
+  - MUST reject the invoice if `invoice_blindedpay` does not contain exactly one `blinded_payinfo` per `invoice_paths`.`blinded_path`.
+- MUST reject the invoice if `features` in any `blinded_payinfo` has any unknown even bits set.
+  - SHOULD confirm authorization if `invoice_request_amount`.`msat` is not within the amount range authorized.
   - if the invoice is a reply to an `invoice_request`:
```
I think this has an extra indentation and below?

rustyrussell replied:
Actually, indentation is wrong on  651 "- MUST reject the invoice if features"... fixed.

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,335 @@ The human-readable prefix for offers is `lno`.
 
 A writer of an offer:
-  - if it sets `node_id`:
+  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 1.
```
```suggestion
  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 0.
```

Think this didn't get updated after `invoice_request_payer_info` became type 0?

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,335 @@ The human-readable prefix for offers is `lno`.
 1. `tlv_stream`: `invoice_request`
 2. types:
-    1. type: 3 (`chain`)
+    1. type: 0 (`invoice_request_payer_info`)
```
IMHO the reasoning this field is 0 rather than in the [80, 160) range is a bit subtly documented, may be nice to add rationale

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
 ## Requirements
 
 A writer of an invoice:
```
I wonder if we should add an upper bound to the TLV space an invoice can use? Guess it seems unlikely we'd want more than 40 required fields in the offer or `inv_req`, but just thought I'd ask

rustyrussell replied:
I always struggle a bit with meta-discussion inside specs.  They're a complete distraction for anyone except someone trying to modify the spec...

If we get to that, we can add an ugly bit extending the range but let's not get there...

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
-    - otherwise:
-      - MUST NOT set `refund_signature`
+      - MUST NOT set `invoice_request_quantity`
+    - MUST set `invoice_request_payer_key` to `offer_node_id`.
```
Curious the reasoning for not just disallowing this field in `offer_send_invoice` invoices?

rustyrussell replied:
Fixed in #11

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
-  - otherwise: (not an `invoice_request` reply, and no `offer_id`):
-    - if `chain` is not present:
+      - MUST reject the invoice if there is an `invoice_request_quantity` field.
+  - otherwise: (not an `invoice_request` reply, nor for `offer_send_invoice`):
```
Hm, this is a weird case. Feels like we should either disallow/fail or also require that `invoice_request_quantity`, etc are set properly.

rustyrussell replied:
Yes, I think the "raw invoice" case is under-considered.  Like, what is allowed in this case?  I think you're right, let's create a checklist of what makes sense here...

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
+  - otherwise (responding to an `invoice_request`):
+     - MUST copy all non-signature fields from the invoice_request (including unknown fields).
+  - MUST set `invoice_created_at` to the number of seconds since Midnight 1
     January 1970, UTC when the offer was created.
```
Should this be the invoice?

rustyrussell replied:
I think so?

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
+  - MUST set `invoice_created_at` to the number of seconds since Midnight 1
     January 1970, UTC when the offer was created.
-  - MUST set `payment_hash` to the SHA2 256-bit hash of the
+  - MUST set `invoice_payment_hash` to the SHA2 256-bit hash of the
```
s/SHA2 256/SHA256?

rustyrussell replied:
Yeah, technically it's 256-bit SHA2, but we use SHA256 everywhere else...

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -551,144 +585,100 @@ using the `onion_message` `invoice` field.
-      seconds after `created_at` that payment of this invoice should not be attempted.
-  - if the `min_final_cltv_expiry` for the last HTLC in the route is not 18:
-    - MUST set `min_final_cltv_expiry`.
+    - MUST set `sig` to the signature using `offer_node_id` as described in [Signature Calculation](#signature-calculation).
```
I think this can only apply if we're responding to an `invoice_request`. If we're sending to a `send_invoice` offer, we don't have the `offer_node_id` privkey to sign with?

rustyrussell replied:
Yes, this is fixed in the https://github.com/rustyrussell/lightning-rfc/pull/11

rustyrussell committed:
Even more feedback from @thomash-acinq and @valentinewallace

1. Remove the "offer_node_id is optional".  It's messy.  Put it in
   even if it's a temporary key.
2. Note why we use 0 for the invoice_request_payer_info.
3. Use SHA256 instead of 256-bit SHA-2, as everyone knows what SHA256 means
   here; so much that the latter is confusing.
4. Indent fix.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: add `invoice_amount` explicitly.

@thomash-acinq points out that changing invoice_request_amount is a
hack.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: make it clear that unhashed elements are 240-1000 *inclusive*.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
     2. data:
         * [`tu64`:`timestamp`]
-    1. type: 42 (`payment_hash`)
+    1. type: 168 (`invoice_payment_hash`)
```
Looks like 166 was skipped.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
     2. data:
         * [`sha256`:`payment_hash`]
-    1. type: 44 (`relative_expiry`)
+    1. type: 170 (`invoice_relative_expiry`)
```
Could we move this immediately after `invoice_created_at`? Seems more natural there.

rustyrussell replied:
OK, shuffled that around, used 166 which was omitted somehow.

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
 1. `tlv_stream`: `invoice_request`
 2. types:
-    1. type: 3 (`chain`)
+    1. type: 0 (`invoice_request_payer_info`)
```
IIUC, I think a better name would be `invoice_request_metadata`, for uniformity with `offer_metadata`. They seem to serve a similar purpose?

rustyrussell replied:
Renamed.

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
 
 Invoice Requests are a request for an invoice; the human-readable prefix for
-invoices is `lnr`.
+invoices is `lnr`.  It mirrors all the fields from the offer, except
```
Seems the current spec doesn't support custom TLVs in `offer`s/`invoice(_req)`s? May need to add a type range for each message's custom tlvs to be in. Otherwise we can't tell whether an inbound custom TLV is being reflected back to us or was added by the counterparty, IIUC.

rustyrussell replied:
As written, custom things would *have* to be in the corresponding ranges if you want them to be sent back to you.

If we want more room, we can either:
1. Add more ranges, eg. 1000-1999, 2000-2999, 3000-3999.
2. Do generic interleaving "if > 1000, mod 6 (0/1 == offer, 2/3 == invreq, 4/5 == invoice)".

We could do generic interleaving on these fields, too (damn, renumber again!).

valentinewallace replied:
> Add more ranges, eg. 1000-1999, 2000-2999, 3000-3999.

ACK on this 👍 the other ones sound a bit weird but I'm fine as long as there's some dedicated space somewhere

carlaKC commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
-    - If so, it MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
-  - MUST set `description` to a complete description of the purpose
+  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 0.
+  - MUST set `offer_node_id` to the node's public key to request the invoice from.
```
Or the final blinded node id if `offer_paths` are specified?

rustyrussell replied:
Unf (as I keep forgetting!) blinded paths are made *forwards*, so you can't control the final node_id.  If you have multiple, the finals will not be the same!

But yeah, if you're using a blinded path, this may not be a *useful* "node's public key".

carlaKC commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
-  - MAY specify exactly one signature TLV: `signature`:
-    - If so, it MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
-  - MUST set `description` to a complete description of the purpose
+  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 0.
```
I think it's worth adding a high level explanation somewhere explaining why we limit the range of TLV fields? 

We still specify `tlv_stream: offer/invoice_request/invoice` as separate streams, but they're actually related which is different to other TLV streams in the spec.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -238,296 +233,328 @@ The human-readable prefix for offers is `lno`.
+  - SHOULD set `invoice_request_payer_info` to an unpredictable series of bytes.
+  - MUST set `invoice_request_payer_key` to a transient public key.
+  - MUST remember the secret key corresponding to `invoice_request_payer_key`.
+  - MUST NOT set or imply any `invoice_request_chain` not set or implied by `offer_chains`.
```
Unclear if not setting `invoice_request_chain ` implies (a) the chain set or implied by `offer_chains` or (b) bitcoin.

rustyrussell replied:
Erk, yeach, this is a pain to specify :(

I've changed it to this:
```
  - if `offer_chains` is set:
    - MUST set `invoice_request_chain` to one of `offer_chains` unless that chain is bitcoin, in which case it MAY omit `invoice_request_chain`.
  - otherwise:
    - if it sets `invoice_request_chain` it MUST set it to bitcoin.
```

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -551,144 +578,102 @@ using the `onion_message` `invoice` field.
-  - for the bitcoin chain, if the invoice specifies `fallbacks`:
+       - MUST reject the invoice if `invoice_request_chain` is not a supported chain.
+  - for the bitcoin chain, if the invoice specifies `invoice_fallbacks`:
     - MUST ignore any `fallback_address` for which `version` is greater than 16.
     - MUST ignore any `fallback_address` for which `address` is less than 2 or greater than 40 bytes.
     - MUST ignore any `fallback_address` for which `address` does not meet known requirements for the given `version`
```
Is there a reason for the more soft "ignore" rather than "reject" if the writer MUST write a valid witness version and program?

rustyrussell replied:
Future expansion...

rustyrussell committed:
BOLT 12: even more feedback!

* Make a note about TLV ranges.
* rename invoice_request_payer_info to invoice_request_metadata.
* Make invoice_request_chain's "bitcoin default" explicit.
* Renumber to place invoice_relative_expiry next to invoice_created_at, close gap in numbering.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
